probability/windowed_median.cpp

```diff
@@ -0,0 +1,189 @@
+/**
+ * @file
+ * @brief An implementation of a median calculation of a sliding window along a data stream
+ *
+ * @details
+ * Given a stream of integers, the algorithm calculates the median of a fix size window at the back of the stream. The leading time complexity of this algorithm is
+ * O(log(N), and it is inspired by the known algorithm to [find median from (infinite) data stream](https://www.tutorialcup.com/interview/algorithm/find-median-from-data-stream.htm), with the proper modifications to account
+ * for the finite window size for which the median is requested
+ *
+ * ### Algorithm
+ * The sliding window is managed by a list, which guarantees O(1) for both pushing and popping. Each new value is pushed to the window back, while a value
+ * from the front of the window is popped. In addition, the algorithm manages a multi-value binary search tree (BST), implemented by std::multiset. For each new
+ * value that is inserted to the window, it is also inserted to the BST. When a value is popped from the window, it is also erased from the BST. Both insertion and
+ * erasion to/from the BST are O(logN) in time, with N the size of the window. Finally, the algorithm keeps a pointer to the root of the BST, and updates its position
+ * whenever values are inserted or erased to/from BST. The root of the tree is the median! Hence, median retrieval is always O(1)
+ *
+ * Time complexity: O(logN). Space complexity: O(N). N - size of window
+ * @author [Yaniv Hollander](https://github.com/YanivHollander)
+ */
+#include <cassert>      /// for assert
+#include <list>         /// for std::list - used to manage sliding window
+#include <set>          /// for std::multiset - used to manage multi-value sorted sliding window values
+
+#include <ctime>        /// for std::time - needed in testing
+#include <cstdlib>      /// for std::rand - needed in testing
+#include <vector>       /// for std::vector - needed in testing
+
+/**
+ * @namespace probability
+ * @brief Probability algorithms
+ */
+namespace probability {
+using Window = std::list<int>;
```

Comment:
Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). **Let us know if you need any help here or in our [Discord server](https://discord.gg/c7MnfGFGa6).** 🙂

Reply:
Are you fine with my suggestion to use `std::list<int>::size_type` as the type for the class member `_windowSize`? It's basically the same as `uint64_t`, but more transparent and portable

Reply:
> Are you fine with my suggestion to use `std::list<int>::size_type` as the type for the class member `_windowSize`? It's basically the same as `uint64_t`, but more transparent and portable

If you think that's better, then let's go with that. 🙂

Reply:
OK